Realm/RLMQueryUtil.mm

```diff
@@ -538,6 +519,43 @@ void add_constraint_to_query(realm::Query &query, RLMPropertyType type,
     }
 }
 
+void do_add_constraint_to_query(realm::Query&, RLMPropertyType, NSPredicateOperatorType,
+                                NSComparisonPredicateOptions, const std::vector<NSUInteger>&,
+                                id, realm::null)
+{
+    // This is not actually reachable as this case is caught earlier, but this
+    // overload is needed for the code to compile
+    @throw RLMPredicateException(@"Invalid predicate expressions",
+                                 @"Predicate expressions must compare a keypath and another keypath or a constant value");
+}
+
+bool is_nsnull(id value) {
+    return !value || value == NSNull.null;
+}
+
+template<typename T>
+bool is_nsnull(T) {
+    return false;
+}
+
+template <typename L, typename R>
+void add_constraint_to_query(realm::Query &query, RLMPropertyType type,
+                             NSPredicateOperatorType operatorType,
+                             NSComparisonPredicateOptions predicateOptions,
+                             const std::vector<NSUInteger>& linkColumns, L lhs, R rhs)
+{
+    // The expression operators are only overloaded for realm::null on the rhs
+    RLMPrecondition(!is_nsnull(lhs), @"Unsupported operator",
+                    @"Nil is only supported on the right side of operators");
+
+    if (is_nsnull(rhs)) {
```

Comment:
this could be simplified a bit by using `auto validatedRHS = is_nsnull(rhs) ? realm::null() : rhs;` and passing the validated value to `do_add_constraint_to_query`.